File: src/Calculator/Converters/ExpressionItemTemplateSelector.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "ExpressionItemTemplateSelector.h"
#include "CalcViewModel/Common/DisplayExpressionToken.h"

using namespace CalculatorApp::Common;

namespace CalculatorApp
{
    namespace Converters
    {
        Windows::UI::Xaml::DataTemplate^ ExpressionItemTemplateSelector::SelectTemplateCore(Platform::Object^ item, Windows::UI::Xaml::DependencyObject^ container)
        {
            DisplayExpressionToken^ token = dynamic_cast<DisplayExpressionToken^>(item);
            if (token != nullptr)
            {
                Common::TokenType type = token->Type;

                switch (type)
                {
                case TokenType::Operator:
                    return m_operatorTemplate;
                case TokenType::Operand:
                    return m_operandTemplate;
                case TokenType::Separator:
                    return m_separatorTemplate;
                default:
                    throw ref new Platform::Exception(E_FAIL, L"Invalid token type");
                }
            }

            return m_separatorTemplate;
        }
    }
}

File: src/Calculator/Converters/ExpressionItemTemplateSelector.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using CalculatorApp.ViewModel.Common;

using System;

using Windows.UI.Xaml;

namespace CalculatorApp
{
    namespace Converters
    {
        [Windows.UI.Xaml.Data.Bindable]
        public sealed class ExpressionItemTemplateSelector : Windows.UI.Xaml.Controls.DataTemplateSelector
        {
            protected override DataTemplate SelectTemplateCore(object item, DependencyObject container)
            {
                DisplayExpressionToken token = (item as DisplayExpressionToken);
                if (token != null)
                {
                    CalculatorApp.ViewModel.Common.TokenType type = token.Type;

                    switch (type)
                    {
                        case TokenType.Operator:
                            return OperatorTemplate;
                        case TokenType.Operand:
                            return OperandTemplate;
                        case TokenType.Separator:
                            return SeparatorTemplate;
                        default:
                            throw new Exception("Invalid token type");
                    }
                }

                return SeparatorTemplate;
            }

            public Windows.UI.Xaml.DataTemplate OperatorTemplate { get; set; }

            public Windows.UI.Xaml.DataTemplate OperandTemplate { get; set; }

            public Windows.UI.Xaml.DataTemplate SeparatorTemplate { get; set; }
        }
    }
}

